### PR TITLE
Screenshot tests

### LINKED ISSUE #9: Set up screenshot tests
We would like to use snapshots for our UI tests to speed up repetitive tasks and better resolution.

It seems that the 2 main contenders are:
* https://github.com/pointfreeco/swift-snapshot-testing
* https://github.com/uber/ios-snapshot-test-case

We should play around with both and see which one we like better.

We also need to find a solution for restoring artefacts between successive runs so we have something to compare with.

It's still up for discussion whether this should be done on Github Actions or Xcode Cloud, mainly because we don't have access to Xcode Cloud yet.

We also talked about adding debug borders around all views to improve snapshot accuracy and, if speedy enough, we can use snapshots within the integration tests as well.

### PR DESCRIPTION
**Pull request by ismailgulek**
Provides a screenshot testing using https://github.com/pointfreeco/swift-snapshot-testing library. I've configured the tests not to fail for newly-taken screenshots.

Tests run on the matrix of:
|Devices|Languages|
|-|-|
|iPhone 13 Pro Max, iPad(9th generation)|en-GB, fr-FR, de-DE|

Fixes #9 

Requires Xcodegen 2.29.0+

### CODE REVIEW

File: UITests/Sources/AuthenticationCoordinatorUITests.swift [hunk 1]
```diff
@@ -44,7 +46,13 @@ class AuthenticationCoordinatorUITests: XCTestCase {
...
         app.textFields["usernameTextField"].tap()
         app.typeText("alice\n")
-        app.typeText("87654321\n")
+        app.secureTextFields["passwordTextField"].tap()
+        app.typeText("87654321")
+
+        // Login Screen: Tap next
+        app.buttons["nextButton"].tap()
+
+        try await Task.sleep(nanoseconds: 1_000_000_000)
```
Comment (pixlwave) [inline UITests/Sources/AuthenticationCoordinatorUITests.swift:55]: Could we perhaps replace this with a `app.alerts.firstMatch.waitForExistence(timeout: 1)`?

File: UITests/SupportingFiles/UITests.xctestplan [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      {
+        "key" : "SNAPSHOT_ARTIFACTS",
+        "value" : "\/tmp\/__FailedScreenshots__"
+      }
+    ],
+    "locationScenario" : {
+      "identifier" : "London, England",
+      "referenceType" : "built-in"
+    },
+    "mainThreadCheckerEnabled" : false,
```
Comment (pixlwave) [inline UITests/SupportingFiles/UITests.xctestplan:42]: How come this is disabled (here and elsewhere)?

File: ElementX/Sources/Other/SwiftUI/Animation/ElementAnimations.swift [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+//  Copyright © 2022 Element. All rights reserved.
+//
+
+import Foundation
+import SwiftUI
+
+extension Animation {
+    /// Disabled if running UI tests, otherwise `Animation.default`
+    static var elementDefault: Animation {
+        ProcessInfo.processInfo.environment["IS_RUNNING_UI_TESTS"] == "1" ? .linear(duration: 0) : .default
```
Comment (pixlwave) [inline ElementX/Sources/Other/SwiftUI/Animation/ElementAnimations.swift:15]: This feels rather fragile. If someone wanted to use `.easeIn(duration:)` or `.interpolatingSpring(mass:stiffness:damping:initialVelocity:)` then we would have to make one of these for each of those uses? 🤔

(The splash screen has a few uses of `.easeInOut` btw).

**Review by pixlwave — COMMENTED** [review]
Amazing to see this working, can't wait to use it! 🥳

I've added some comments inline, one thing I haven't added though is that I don't think we should use python 2 for this and should probably switch to python 3 as its more likely someone has installed that.

File: UITests/Sources/Application.swift [hunk 4]
```diff
@@ -28,4 +30,45 @@ extension XCUIApplication {
...
+    }
+
+    private var osVersion: String {
+        UIDevice.current.systemVersion.replacingOccurrences(of: ".", with: "-")
+    }
+}
+
+public extension Snapshotting where Value == XCUIElement, Format == UIImage {
+    // swiftformat:disable:next redundanttype
+    static let image: Snapshotting =
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply UITests/Sources/Application.swift:72]: This is actually not used anymore, removed.

File: UITests/Sources/SessionVerificationUITests.swift [hunk 5]
```diff
@@ -22,11 +22,13 @@ class SessionVerificationUITests: XCTestCase {
...
         app.goToScreenWithIdentifier(.sessionVerification)
         
-        XCTAssert(app.navigationBars["Verify this session"].exists)
+        XCTAssert(app.navigationBars[ElementL10n.verificationVerifyDevice].exists)
         
         XCTAssert(app.buttons["startButton"].exists)
         XCTAssert(app.buttons["dismissButton"].exists)
         XCTAssert(app.staticTexts["titleLabel"].exists)
+
+        app.assertScreenshot(.sessionVerification)
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply UITests/Sources/SessionVerificationUITests.swift:31]: Or we could define new screen identifiers to achieve that. Let's discuss it in the next round.

File: ElementX/Sources/UITestsRootView.swift [hunk 6]
```diff
@@ -23,5 +23,6 @@ struct UITestsRootView: View {
             .listStyle(.plain)
         }
         .navigationTitle("Screens")
+        .navigationViewStyle(StackNavigationViewStyle())
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply ElementX/Sources/UITestsRootView.swift:26]: 👍

File: UnitTests/Sources/BugReportViewModelTests.swift [hunk 7]
```diff
@@ -44,6 +44,7 @@ class BugReportViewModelTests: XCTestCase {
 
         context.send(viewAction: .removeScreenshot)
         await Task.yield()
+        try await Task.sleep(nanoseconds: 100_000_000)
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply UnitTests/Sources/BugReportViewModelTests.swift:47]: Yeah, let's see if it works.

File: ElementX/Sources/Screens/SplashScreen/SplashScreenModels.swift [hunk 8]
```diff
@@ -53,6 +53,9 @@ struct SplashScreenViewState: BindableState {
     
     /// The background gradient for all 4 pages and the hidden page at the start of the carousel.
     var backgroundGradient: Gradient {
+        if ProcessInfo.processInfo.environment["IS_RUNNING_UI_TESTS"] == "1" {
+            return Gradient(colors: [.white])
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply ElementX/Sources/Screens/SplashScreen/SplashScreenModels.swift:57]: Actually i wouldn't expect that but it's :)

File: Tools/Scripts/get_simulator_time.rb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+require 'time'
+
+time = Time.new(2007, 1, 9, 9, 41, 0)
+print(time.iso8601)
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply Tools/Scripts/get_simulator_time.rb:4]: Well, it was very complicated in Python 2.7. Yeah it's fixing the timezone issue (actually i've copied this from `fastlane snapshot` library). We can probably remove it if we choose to use Python 3+.

File: UITests/Sources/AuthenticationCoordinatorUITests.swift (hunk 1 above)
Comment (ismailgulek) [reply UITests/Sources/AuthenticationCoordinatorUITests.swift:55]: Makes sense, updated.

File: UITests/SupportingFiles/UITests.xctestplan (hunk 2 above)
Comment (ismailgulek) [reply UITests/SupportingFiles/UITests.xctestplan:42]: Well, i've copied the value from El-iOS actually. But yeah that doesn't make sense to disable it, updated.

File: ElementX/Sources/Other/Tests.swift [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+//  Tests.swift
+//  ElementX
+//
+//  Created by Ismail on 29.07.2022.
+//  Copyright © 2022 Element. All rights reserved.
+//
+
+import Foundation
+
+public enum Tests {
```
Comment (pixlwave) [inline ElementX/Sources/Other/Tests.swift:11]: 👌

File: UITests/Sources/SessionVerificationUITests.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (pixlwave) [reply UITests/Sources/SessionVerificationUITests.swift:31]: Makes sense to me 👍

**Review by pixlwave — APPROVED** [review]
Looks great to me! Thanks for the changes to use python3 🙏

**Pull request merged**